File: src/bots/fix_pt_months.php
<?php

namespace WpRefs\FixPtMonth;
/*
usage:

use function WpRefs\FixPtMonth\pt_months;

*/

// use function WikiParse\Citations\get_full_refs;
use function WikiParse\Citations\getCitations;
use function WikiParse\Template\getTemplate;
use function WikiParse\Template\getTemplates;
use function WpRefs\TestBot\echo_test;
use function WpRefs\TestBot\echo_debug;
use function WpRefs\Bots\MonthNewValue\make_date_new_val_pt;

function fix_one_cite_text($temp_text)
{
    // ---
    $temp_text = trim($temp_text);
    // ---
    $temp = getTemplate($temp_text);
    // ---
    $params = $temp->getParameters();
    // ---
    foreach ($params as $key => $value) {
        // ---
        $new_value = make_date_new_val_pt($value);
        // ---
        if ($new_value && $new_value != $value) {
            $temp->setParameter($key, $new_value);
        }
    }
    // ---
    $new_text = $temp->toString();
    // ---
    return $new_text;
}


function fix_cites_text($temp_text)
{
    // ---
    $new_text = $temp_text;
    // ---
    $temp_text = trim($temp_text);
    // ---
    $temps = getTemplates($temp_text);
    // ---
    foreach ($temps as $temp) {
        // ---
        $temp_old = $temp->getOriginalText();
        // ---
        // echo_debug("temp_old:($temp_old)\n");
        // ---
        $params = $temp->getParameters();
        // ---
        foreach ($params as $key => $value) {
            // ---
            $new_value = make_date_new_val_pt($value);
            // ---
            if ($new_value && $new_value != $value) {
                $temp->setParameter($key, $new_value);
            }
        }
        // ---
        $temp_new = $temp->toString();
        // ---
        $new_text = str_replace($temp_old, $temp_new, $new_text);
        // ---
    }
    return $new_text;
}


function start_end($cite_temp)
{
    return strpos($cite_temp, "{{") === 0 && strrpos($cite_temp, "}}") === strlen($cite_temp) - 2;
}

function pt_months($text)
{
    // ---
    // echo_test("pt_months:");
    // ---
    $citations = getCitations($text);
    // ---
    $new_text = $text;
    // ---
    foreach ($citations as $key => $citation) {
        // ---
        $cite_temp = $citation->getContent();
        // ---
        // echo_debug($cite_temp);
        // ---
        // if $cite_temp startwith {{ and ends with }}
        // if (start_end($cite_temp) || defined("DEBUG") || True) {
        // ---
        // echo_debug("\n$cite_temp\n");
        // ---
        $new_temp = fix_cites_text($cite_temp);
        // $new_temp = fix_one_cite_text($cite_temp);
        // ---
        // if ($new_temp != $cite_temp) echo_debug("new_temp != cite_temp\n");
        // ---
        $new_text = str_replace($cite_temp, $new_temp, $new_text);
        // } else {
        //     // ---
        //     echo_debug("temp not okay: $cite_temp\n");
        // }
    }
    // ---
    return $new_text;
}

function rm_ref_spaces($newtext)
{
    // ---
    // \s*(\.|,|。|।)\s*((?:\s*<ref[\s\S]+?(?:<\/ref|\/)>)+)
    // ---
    $dot = "(\.|,|。|।)";
    // ---
    $regline = "((?:\s*<ref[\s\S]+?(?:<\/ref|\/)>)+)";
    // ---
    $pattern = "/\s*" . $dot . "\s*" . $regline . "/m";
    $replacement = "$1$2";
    // ---
    $newtext = preg_replace($pattern, $replacement, $newtext);
    // ---
    return $newtext;
}

function pt_fixes($text)
{
    // ---
    $text = pt_months($text);
    // ---
    $text = rm_ref_spaces($text);
    // ---
    return $text;
}

File: python/src/pt_bots/fix_pt_months.py
"""
Portuguese month fixes module

Usage:
    from src.pt_bots.fix_pt_months import pt_fixes
"""

import re
from src.test_bot import echo_test, echo_debug
from src.Parse.Citations import getCitationsOld
from src.WikiParse.Template import getTemplates
from src.bots.months_new_value import make_date_new_val_pt


def start_end(cite_temp):
    """
    Check if text starts with {{ and ends with }}
    
    Args:
        cite_temp: Text to check
        
    Returns:
        True if starts with {{ and ends with }}
    """
    return cite_temp.startswith("{{") and cite_temp.endswith("}}")


def rm_ref_spaces(newtext):
    """
    Remove spaces around punctuation before refs
    
    Args:
        newtext: Text to process
        
    Returns:
        Modified text
    """
    dot = r"(\.|,|。|।)"
    regline = r"((?:\s*<ref[\s\S]+?(?:</ref|/)>)+)"
    pattern = r"\s*" + dot + r"\s*" + regline
    replacement = r"\1\2"
    newtext = re.sub(pattern, replacement, newtext, flags=re.MULTILINE)
    return newtext


def fix_pt_months_in_texts(temp_text):
    """
    Fix Portuguese months in template text
    
    Args:
        temp_text: Text to process
        
    Returns:
        Modified text
    """
    new_text = temp_text
    temp_text = temp_text.strip()
    
    temps = getTemplates(temp_text)
    
    for temp in temps:
        temp_old = temp.getOriginalText()
        
        params = temp.getParameters()
        
        for key, value in params.items():
            new_value = make_date_new_val_pt(value)
            
            if new_value is not None and new_value.strip() != str(value).strip():
                temp.setParameter(key, new_value)
        
        temp_new = temp.toString()
        new_text = new_text.replace(temp_old, temp_new)
    
    return new_text


def fix_pt_months_in_refs(text):
    """
    Fix Portuguese months in references
    
    Args:
        text: Text to process
        
    Returns:
        Modified text
    """
    echo_debug("\n fix_pt_months_in_refs:\n")
    
    new_text = text
    
    citations = getCitationsOld(text)
    
    for citation in citations:
        cite_temp = citation.getContent()
        
        new_temp = fix_pt_months_in_texts(cite_temp)
        
        new_text = new_text.replace(cite_temp, new_temp)
    
    return new_text


def pt_fixes(text):
    """
    Apply Portuguese-specific fixes
    
    Args:
        text: Text to process
        
    Returns:
        Modified text
    """
    text = fix_pt_months_in_refs(text)
    text = rm_ref_spaces(text)
    return text
